peps/pep-0667.rst

```diff
@@ -40,72 +62,227 @@ The current implementation of ``locals()`` and ``frame.f_locals`` is slow,
+
+    def f():
+        exec("x = 1")
+        print(locals().get("x"))
+        x = 0
+    f()
+
+prints ``None`` (the default value from the ``.get()`` call).
+
+With this PEP both examples would print ``None``, as the call to
+``exec()`` and the subsequent call to ``locals()`` would use
+independent dictionary snapshots of the local variables rather
+than using the same shared dictionary cached on the frame object.
+
 Rationale
 =========
 
+Making the ``frame.f_locals`` attribute a write-through proxy
+-------------------------------------------------------------
+
 The current implementation of ``frame.f_locals`` returns a dictionary
-that is created on the fly from the array of local variables.
+that is created on the fly from the array of local variables. The
+``PyFrame_LocalsToFast()`` C API is then called by debuggers and trace
+functions that want to write their changes back to the array (until
+Python 3.11, this API was called implicitly after every trace function
+invocation rather than being called explicitly by the trace functions).
+
 This can result in the array and dictionary getting out of sync with
-each other. Writes to the ``f_locals`` may not show up as
-modifications to local variables. Writes to local variables can get lost.
+each other. Writes to the ``f_locals`` frame atribute may not show up as
+modifications to local variables if ``PyFrame_LocalsToFast()`` is never
+called. Writes to local variables can get lost if the dictionary snapshot
+is not refreshed before being written back to the frame.
 
 By making ``frame.f_locals`` return a view on the
 underlying frame, these problems go away. ``frame.f_locals`` is always in
 sync with the frame because it is a view of it, not a copy of it.
 
+Making the ``locals()`` builtin return independent snapshots
+------------------------------------------------------------
+
+When ``exec`` was converted from a statement to a builtin function
+in Python 3.0 (part of the core language changes in :pep:`3100`), the
+associated implicit call to ``PyFrame_LocalsToFast()`` was removed, so
+it typically appears as if attempts to write to local variables with
+``exec`` in optimized frames are ignored::
+
+    >>> def f():
+    ...     x = 0
+    ...     exec("x = 1")
+    ...     print(x)
+    ...     print(locals()["x"])
+    ...
+    >>> f()
+    0
+    0
+
+In truth, the writes aren't being ignored, they just aren't
+being copied from the dictionary cache back to the optimized local
+variable array. The changes to the dictonary are then overwritten
+the next time the dictionary cache is refreshed from the array::
+
+    >>> def f():
+    ...     x = 0
+    ...     locals_cache = locals()
+    ...     exec("x = 1")
+    ...     print(x)
+    ...     print(locals_cache["x"])
+    ...     print(locals()["x"])
+    ...
+    >>> f()
+    0
+    1
+    0
+
+The behaviour becomes even stranger if a tracing function
+or another piece of code invokes ``PyLocals_ToFast()`` before
```

Comment:
Is `PyLocals_ToFast()` a thing?

Reply:
Nope, I just have a mental shorthand for `PyFrame_LocalsToFast()` that sometimes escapes through my fingers when typing.